Adds the AudioData interface (#8511)
mdn/content: files/en-us/web/api/audiodata/index.md

browser-compat: api.AudioData
---
{{DefaultAPISidebar("WebCodecs API")}}

The **`AudioData`** interface of the {{domxref('WebCodecs API')}} represents an audio sample.

## Description

An audio track consists of a stream of audio samples, each sample representing a captured moment of sound. An `AudioData` object is a representation of such a sample. Working alongside the interfaces of the [Insertable Streams API](/en-US/docs/Web/API/Insertable_Streams_for_MediaStreamTrack_API), you can break a stream into individual `AudioData` objects with {{domxref("MediaStreamTrackProcessor")}}, or construct an audio track from a stream of frames with {{domxref("MediaStreamTrackGenerator")}}.

> **Note**: Find out more about audio on the web in [Digital audio concepts](/en-US/docs/Web/Media/Formats/Audio_concepts).

### The media resource

An `AudioData` object contains a reference to an attached **media resource**. This media resource contains the actual audio sample data described by the object. A media resource is maintained by the user agent until it is no longer referenced by an `AudioData` object, for example when {{domxref("AudioData.close()")}} is called.

### Planes and audio format

> To return the sample format of an `AudioData` use the {{domxref("AudioData.format")}} property. The format may be described as **interleaved** or **planar**. In a interleaved format, the audio samples are laid out in a single buffer, therefore the `AudioData` has a single **plane**. This plane contains a number of elements equal to {{domxref("AudioData.numberOfFrames")}} * {{domxref("AudioData.numberOfChannels")}}.

Reviewer: ```suggestion
To return the sample format of an `AudioData` object use the {{domxref("AudioData.format")}} property. The format may be described as **interleaved** or **planar**. In an interleaved format, the audio samples are laid out in a single buffer, therefore the `AudioData` has a single **plane**. This plane contains a number of elements equal to {{domxref("AudioData.numberOfFrames")}} * {{domxref("AudioData.numberOfChannels")}}.
```

Reviewer: >In an interleaved format, the audio samples are laid out in a single buffer, therefore the `AudioData` has a single **plane**.

This latter suggests interleaved and planer, yet you said interleaved _or_ planer earlier. What am I missing?

Reviewer: Closing this in favor of a new suggestin.

> To return the sample format of an `AudioData` use the {{domxref("AudioData.format")}} property. The format may be described as **interleaved** or **planar**. In a interleaved format, the audio samples are laid out in a single buffer, therefore the `AudioData` has a single **plane**. This plane contains a number of elements equal to {{domxref("AudioData.numberOfFrames")}} * {{domxref("AudioData.numberOfChannels")}}.

Reviewer: ```suggestion
To return the sample format of an `AudioData` object use the {{domxref("AudioData.format")}} property. The format is stored in buffers called planes, and may be described as **interleaved** or **planar**. In interleaved format, the audio samples are a single plane that contains a number of elements equal to {{domxref("AudioData.numberOfFrames")}} times {{domxref("AudioData.numberOfChannels")}}.
```

Reviewer: >The format is stored in buffers called planes

This is the keystone of my proposal. It allows later descriptions to be simplified.

Reviewer: These two suggestions rise and fall together. 

I don't know if I've improved this merely substituted my writing preferences for yours. Tell me what you think.

Reviewer: Thank you for your patience.

Author: My only concern is that the spec does not state that buffers are called planes, I'm wondering if by simplifying this we actually infer some relationship that isn't quite correct (though may be correct in terms of how WebCodecs deals with it). I don't know enough about audio to be able to say. I think my version maintains the distinction in the spec, yours makes them equivalent. If that is correct, then the simplification works.

Reviewer: If planes aren't buffers, then I don't understand the relationship between them and I can't judge whether your version is even correct. Please ping the spec author for clarification.

Reviewer 2: I'm using buffer as synonym for "array". Buffer is a more generic concept than plane. It would be correct to say that any plane is a buffer, but not the other way around. 

For example, if you called

`audioData.copyTo(destination, {planeIndex: 0})`

After this call I would describe `destination` as being a buffer containing (a copy of) plane 0. 

If you instead called

`audioData.copyTo(destination, {planeIndex: 0, frameCount: 1})`

I'd still call `destination` a buffer, but now it only contains a single frame from plane 0.

I read both of your phrasing as being correct w/ some small nits of my own.
- "In a interleaved format.." - the 'a' is helpful here because there are multiple interleaved formats. Alternatively, you could pluralize format.
- "... the audio samples are" - I'd mention channels here as you do for planar format. e.g. "the audio samples _from the different channels_ are".
- "The format is stored in buffers called planes" - little fast/loose, but maybe ok if you were both clear on the meaning. Technically it's the "data" or "samples" that are stored in buffers, with a arrangement defined by "format".  
-

Author: @jpmedley I haven't got back to this yet, elsewhere this also refers to transferable objects which have been removed in the last few days so we had a discussion about that and I raised https://github.com/mdn/content/pull/9141, I figured I could link that it once it is merged when I edit this one.

I'll try and edit this to be correct and acceptable at the same time. :)

Reviewer: ack

Author: @jpmedley latest commit addresses this. I think.

> In a planar format, the audio samples from the different channels are laid out in different buffers. The `AudioData` has a number of planes equal to {{domxref("AudioData.numberOfChannels")}}, each plane contains a number of elements equal to {{domxref("AudioData.numberOfFrames")}}.

Reviewer: What about interleaved format?

Author: As explained in the para above, it has a single plane.

Reviewer: Both definitions mention planes. That's why I'm confused.

Reviewer: Is 'interleaved' simply the definition for a single-plane format?

Author: yes both have planes, interleaved only has one. I'm no expert on audio formats, but the spec text and algorithms treat it as such.

> In a planar format, the audio samples from the different channels are laid out in different buffers. The `AudioData` has a number of planes equal to {{domxref("AudioData.numberOfChannels")}}, each plane contains a number of elements equal to {{domxref("AudioData.numberOfFrames")}}.

Reviewer: ```suggestion
In planar format, the number of planes is equal to {{domxref("AudioData.numberOfChannels")}}, and each plane contains a number of elements equal to {{domxref("AudioData.numberOfFrames")}}.
```

Author: I think this is fine with regard to the spec, whatever the answer to the above point is!

## Constructor

- {{domxref("AudioData.AudioData()")}}
  - : Creates a new `AudioData` object.

## Properties

- {{domxref("AudioData.format")}}{{ReadOnlyInline}}

>   - : Returns the sample format of the audio.

Reviewer: Florian recently challenged me on the the term 'Returns' when referring to properties, stating that it should only be used for methods. I never had a problem with this. What are your thoughts? ( I can't remember what he used instead.)

Reviewer 3: He uses "represents" but he is open to feedback on this.

Author: What was his argument? I think if you asked a dev to describe what is happening they would say that it returns the value.

The Write an API documentation also agrees currently: "The InterfaceName.property read-only property returns a {{domxref("type")}} that..."

- {{domxref("AudioData.sampleRate")}}{{ReadOnlyInline}}
  - : Returns the sample rate of the audio in Hz.
- {{domxref("AudioData.numberofFrames")}}{{ReadOnlyInline}}
  - : Returns the number of frames.
- {{domxref("AudioData.numberofChannels")}}{{ReadOnlyInline}}
  - : Returns the number of audio channels.
- {{domxref("AudioData.duration")}}{{ReadOnlyInline}}
  - : Returns the duration of the audio in microseconds.
- {{domxref("AudioData.timestamp")}}{{ReadOnlyInline}}
  - : Returns the timestamp of the audio in microseconds.

## Methods

- {{domxref("AudioData.allocationSize()")}}
  - : Returns the number of bytes required to hold the sample as filtered by options passed into the method.
- {{domxref("AudioData.copyTo()")}}
  - : Copies the samples from the specified plane of the `AudioData` object to the destination.
- {{domxref("AudioData.clone()")}}
  - : Creates a new `AudioData` object with reference to the same media resource as the original.
- {{domxref("AudioData.close()")}}

>   - : Clears all state and releases the reference to the media resource.

Reviewer: ```suggestion
  - : Clears all states and releases the reference to the media resource.
```
